internal/client/consensus/grandpa/communication.go

```diff
@@ -0,0 +1,849 @@
+	gossipEngineMtx sync.Mutex
+	validator       *gossipValidator[H, N, Hasher]
+
+	// Sender side of the neighbor packet channel.
+	// Packets sent into this channel are processed by the `NeighborPacketWorker` and passed on to the underlying
+	// `GossipEngine`.
+	neighborSender neighbourPacketSender[N]
+
+	// `NeighborPacketWorker` processing packets sent through the `NeighborPacketSender`.
+	neighborPacketWorker    neighborPacketWorker[N]
+	neighborPacketWorkerMtx sync.Mutex //nolint: unused
+
+	// Receiver side of the peer report stream populated by the gossip validator, forwarded to the gossip engine.
+	gossipValidatorReportStream    chan peerReport
+	gossipValidatorReportStreamMtx sync.Mutex //nolint: unused
+
+	// TODO: telemetry
+}
+
+func newNetworkBridge[H runtime.Hash, N runtime.Number, Hasher runtime.Hasher[H]](
+	service Network,
+	sync Syncing[H, N],
+	notificationService service.NotificationService,
+	config Config,
+	setState *SharedVoterSetState[H, N],
+	// TODO: metrics, telemetry
+) *networkBridge[H, N, Hasher] {
+	protocol := config.ProtocolName
+	validator, reportStream := newGossipValidator[H, N, Hasher](config, setState)
+
+	gossipEngine := gossip.NewGossipEngine[H, N, Hasher](service, sync, notificationService, protocol, validator)
+
+	{
+		// register all previous votes with the gossip service so that they're available to peers potentially stuck on
+		// a previous round.
+		setState.innerMtx.RLock()
+		completed := setState.inner.completedRounds()
+		setState.innerMtx.RUnlock()
+		setID, voters := completed.setInfo()
+		validator.noteSet(SetID(setID), voters, func(to []peerid.PeerID, msg neighborPacket[N]) {})
+
+		for _, round := range completed.iter() {
+			topic := roundTopic[H, Hasher](Round(round.Number), SetID(setID))
+
+			// we need to note the round with the gossip validator otherwise messages will be ignored.
+			validator.noteRound(Round(round.Number), func(to []peerid.PeerID, msg neighborPacket[N]) {})
+
+			for _, signed := range round.Votes {
+				var gossipMessage gossipMessageVDT[H, N]
+				gossipMessage.inner = gossipMessageVote[H, N](voteMessage[H, N]{
+					Message: signed,
+					Round:   Round(round.Number),
+					SetID:   SetID(setID),
+				})
+				gossipEngine.RegisterGossipMessage(topic, scale.MustMarshal(gossipMessage))
+			}
+
+			logger.Tracef(
+				"Registered %d messages for topic %v (round: %d, setID: %d)",
+				len(round.Votes), topic, round.Number, setID,
+			)
+		}
+	}
+
+	neighborPacketWorker, neighborPacketSender := newNeighborPacketWorker[N](neighborRebroadcastPeriod)
+	nb := networkBridge[H, N, Hasher]{
+		service:                     service,
+		sync:                        sync,
+		gossipEngine:                gossipEngine,
+		validator:                   validator,
+		neighborSender:              neighborPacketSender,
+		neighborPacketWorker:        neighborPacketWorker,
+		gossipValidatorReportStream: reportStream,
+	}
+
+	go func() {
```

Comment:
Should this live in a seperate `run()` function instead of being in the constructor?